pygen/pygen_src/test/riscv_instr_cov_test.py

```diff
@@ -0,0 +1,194 @@
+                # Get the header line
+                for row in csv_reader:
+                    if line_count == 0:
+                        header = row
+                        logging.info("Header: {}".format(header))
+                    else:
+                        entry = row
+                        if len(entry) != len(header):
+                            logging.info("Skipping malformed entry[{}]: "
+                                         "[{}]".format(self.entry_cnt, entry))
+                            self.skipped_cnt += 1
+                        else:
+                            self.trace["csv_entry"] = row
+                            for idx in range(len(header)):
+                                if "illegal" in entry[idx]:
+                                    expect_illegal_instr = True
+                                self.trace[header[idx]] = entry[idx]
+                                if header[idx] != "pad":
+                                    logging.info("{} = {}".format(header[idx],
+                                                                  entry[idx])) 
+                            self.post_process_trace()
+                            if self.trace["instr"] in ["li", "ret", "la"]:
+                                pass
+                            if "amo" in self.trace["instr"] or \
+                               "lr" in self.trace["instr"] or \
+                               "sc" in self.trace["instr"]:
+                                # TODO: Enable functional coverage for AMO test
+                                pass
+                            if not self.sample():
+                                if not expect_illegal_instr:
+                                    logging.error("Found unexpected illegal "
+                                                  "instr: {} "
+                                                  "[{}]".format(self.trace[
+                                                      "instr"],entry))
+                                    self.unexpected_illegal_instr_cnt += 1
+                        self.entry_cnt += 1
+                    line_count += 1
+                logging.info("[{}]: {} instr processed".format(csv_file,
+                                                               self.entry_cnt))
+                self.total_entry_cnt += self.entry_cnt
+        logging.info("Finished processing {} trace CSV, {} "
+                     "instructions".format(len(self.trace_csv),
+                                          self.total_entry_cnt))
+        if self.skipped_cnt > 0 or self.unexpected_illegal_instr_cnt > 0:
+            logging.error("{} instruction skipped, {} illegal "
+                          "instructions".format(self.skipped_cnt),
+                          self.unexpected_illegal_instr_cnt)
+
+    def post_process_trace(self):
+        pass
+
+    def sample(self):
+        instr_name, binary = "", ""
+        get_val(self.trace["binary"], binary, hexa=1)
```

Comment:
maybe we should have a standardized usage of `get_val(...)`? 
Seems like right now the usage depends on the type of variable it returns (e.g. here we pass the `binary` string by reference but below on line 133 we explicitly assigns the return value to `instruction.pc` - maybe we can eliminate the `output_val` argument of `get_val()` and always explicitly assign a return value for clarity?

Reply:
That makes sense. It's fixed.